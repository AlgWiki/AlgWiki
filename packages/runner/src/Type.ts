// The types challenges accept
export enum Primitive {
  // Only supporting signed integers for now
  Integer,
  // 64 bit floating point (IEEE754)
  Float,
  String,
  Boolean,
}

// This is the TypeScript type of the primitive
export type InferredPrimitive<P> = P extends Primitive.Integer | Primitive.Float
  ? number
  : P extends Primitive.String
  ? string
  : P extends Primitive.Boolean
  ? boolean
  : never;

// All Types require this for type discrimination
export interface Variant {
  tag: keyof Renderer;
  type: unknown;
  value?: unknown;
  render: (renderer: Renderer) => string;
}

// A utility interface for easily matching/switching across variants
export interface Renderer {
  single: <P extends Primitive>(t: P, v: InferredPrimitive<P>) => string;
  list: <P extends Primitive>(t: P, v: InferredPrimitive<P>[]) => string;
  linkedList: <P extends Primitive>(t: P, v: InferredPrimitive<P>[]) => string;
  dictionary: <K extends Primitive, V extends Primitive>(
    t: KeyPair<K, V>,
    v: Map<InferredPrimitive<K>, InferredPrimitive<V>>
  ) => string;
}

function throwIfNeedsValue<V extends Variant>(
  v: V,
  f: Renderer[V["tag"]]
): string {
  if (f.length == 2 && !v.value) {
    throw new Error(
      `Attempted to render a type-only template with a value: ${v}`
    );
  }

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  return f(v.type, v.value) as string;
}

// Just a single value
class Single<P extends Primitive> implements Variant {
  public readonly tag = "single";
  public constructor(public type: P, public value?: InferredPrimitive<P>) {}
  public render = (r: Renderer): string =>
    throwIfNeedsValue(this, r.single.bind(r));
}

// A list, all the values are of the same type
class List<P extends Primitive> implements Variant {
  public readonly tag = "list";
  public constructor(public type: P, public value?: InferredPrimitive<P>[]) {}
  public render = (r: Renderer): string =>
    throwIfNeedsValue(this, r.list.bind(r));
}

// A linked list, all the values are of the same type
class LinkedList<P extends Primitive> implements Variant {
  public readonly tag = "linkedList";
  public constructor(public type: P, public value?: InferredPrimitive<P>[]) {}
  public render = (r: Renderer): string =>
    throwIfNeedsValue(this, r.linkedList.bind(r));
}

// A dictionary, all the keys are of the same type, and all the values are of the same type
export type KeyPair<K, V> = { key: K; value: V };
class Dictionary<K extends Primitive, V extends Primitive> implements Variant {
  public readonly tag = "dictionary";
  public constructor(
    public type: KeyPair<K, V>,
    public value?: Map<InferredPrimitive<K>, InferredPrimitive<V>>
  ) {}
  public render = (r: Renderer): string =>
    throwIfNeedsValue(this, r.dictionary.bind(r));
}

// Simple wrapper class to make constructing these easier
export class Type<V extends Variant> {
  // public creation api
  public static single<P extends Primitive>(
    inner: P,
    value?: InferredPrimitive<P>
  ): Type<Single<P>> {
    return new Type(new Single(inner, value));
  }

  public static list<P extends Primitive>(
    inner: P,
    value?: InferredPrimitive<P>[]
  ): Type<List<P>> {
    return new Type(new List(inner, value));
  }

  public static linkedList = <P extends Primitive>(
    inner: P,
    value?: InferredPrimitive<P>[]
  ): Type<LinkedList<P>> => {
    return new Type(new LinkedList(inner, value));
  };

  public static dictionary = <K extends Primitive, V extends Primitive>(
    inner: KeyPair<K, V>,
    value?: Map<InferredPrimitive<K>, InferredPrimitive<V>>
  ): Type<Dictionary<K, V>> => {
    return new Type(new Dictionary(inner, value));
  };

  // this is private to force construction via static apis
  private constructor(private readonly inner: V) {}

  public render(renderer: Renderer): string {
    return this.inner.render(renderer);
  }
}
