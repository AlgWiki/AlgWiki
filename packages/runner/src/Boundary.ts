import { randomBytes } from "crypto";
import { promisify } from "util";

export interface IBoundary {
  readonly marker: string;
  readonly error: string;
  readonly start: string;
  readonly end: string;
}

export class Boundary implements IBoundary {
  public readonly error: string;
  public readonly start: string;
  public readonly end: string;

  public static async create(): Promise<Boundary> {
    const buf = await promisify(randomBytes)(16);
    return new Boundary(buf.toString("hex"));
  }

  private constructor(public readonly marker: string) {
    this.error = `${this.marker}_error`;
    this.start = `${this.marker}_start`;
    this.end = `${this.marker}_end`;
  }

  public toString(): string {
    return this.marker;
  }
}
