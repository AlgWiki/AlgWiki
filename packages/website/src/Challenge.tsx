import * as React from "react";

import { Button } from "@material-ui/core";
import CloudUploadIcon from "@material-ui/icons/CloudUpload";

import { CodeEditor } from "./CodeEditor";
import { TestCases } from "./TestCases/TestCases";

const STARTING_CODE = `const fizzbuzz = () => {
  return [1, 2, "Fizz", 4, "Buzz", 6, "etc..."];
};`;

const fizzbuzz = (): string[] =>
  [...Array(99).keys()].map((n) =>
    ++n % 3 === 0
      ? n % 5 === 0
        ? "FizzBuzz"
        : "Fizz"
      : n % 5
      ? "Buzz"
      : `${n}`
  );

export const Challenge: React.FC = () => {
  const [charCount, setCharCount] = React.useState(0);
  const codeRef = React.useRef<() => string>();
  return (
    <div>
      <h2>FizzBuzz</h2>
      <article>
        Output a list of numbers from 1 to 99 (including 1 and 99), but:
        <ul>
          <li>
            If the number is divisble by 3, output <code>Fizz</code>
          </li>
          <li>
            If the number is divisble by 5, output <code>Buzz</code>
          </li>
          <li>
            If the number is divisble by both 3 and 5, output{" "}
            <code>FizzBuzz</code>
          </li>
          You may output the numbers as strings or numbers.
        </ul>
      </article>
      <div>Encoding: ASCII, Bytes: {charCount}</div>
      <CodeEditor
        initialCode={STARTING_CODE}
        onChange={setCharCount}
        getCodeRef={codeRef}
      />
      <TestCases
        functionName="fibonacci"
        testCases={[{ input: [], expected: fizzbuzz() }]}
      />
      <Button
        variant="contained"
        color="primary"
        startIcon={<CloudUploadIcon />}
      >
        Submit
      </Button>
    </div>
  );
};
