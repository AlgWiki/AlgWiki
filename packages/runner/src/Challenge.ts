import { Boundary } from "./Boundary";
import { Type, Variant } from "./Type";

export class Challenge<Input extends Variant, Output extends Variant> {
  public constructor(
    public readonly name: string,
    public readonly inputs: Type<Input>[],
    public readonly output: Type<Output>
  ) {}

  public createDefaultCode(renderer: ChallengeRenderer<Input, Output>): string {
    return renderer.createDefaultCode(this);
  }

  public async createRunner(
    renderer: ChallengeRenderer<Input, Output>,
    boundary: Boundary,
    userCode: string
  ): Promise<string> {
    return renderer.createRunner(this, boundary, userCode);
  }

  public inputJsonString(): string {
    return JSON.stringify(
      this.inputs.map((inp) => {
        if (!inp.inner.value) {
          throw new Error(
            "tried to serialise Type to JSON, but had no inner value!"
          );
        }

        if (inp.isDictionary()) {
          const inner = inp.inner.value as Map<unknown, unknown>;
          return [...inner.entries()].map(([k, v]) => [k, v]);
        }

        return inp.inner.value;
      })
    );
  }
}

export interface ChallengeRenderer<
  Input extends Variant,
  Output extends Variant
> {
  // The base template for the user code
  createDefaultCode(challenge: Challenge<Input, Output>): string;
  // The outputted template for the runner
  createRunner(
    challenge: Challenge<Input, Output>,
    boundary: Boundary,
    userCode: string
  ): string;
}
