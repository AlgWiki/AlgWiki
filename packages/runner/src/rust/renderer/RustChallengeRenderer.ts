import { snakeCase } from "change-case";
import dedent from "dedent";

import { IBoundary } from "../../Boundary";
import { Challenge, ChallengeRenderer } from "../../Challenge";
import { Dictionary, LinkedList, Type, Variant } from "../../Type";
import { RustTypeRenderer } from "./RustTypeRenderer";
import { RustValueRenderer } from "./RustValueRenderer";

export class RustChallengeRenderer<
  Input extends Variant,
  Output extends Variant
> implements ChallengeRenderer<Input, Output>
{
  private readonly typeRenderer = new RustTypeRenderer();
  private readonly valueRenderer = new RustValueRenderer();

  private identifier(name: string): string {
    return snakeCase(name);
  }

  public createDefaultCode({
    name,
    inputs,
    output,
  }: Challenge<Input, Output>): string {
    const challengeIdent = this.identifier(name);
    const inputTypeStr = inputs[0].render(this.typeRenderer);
    const outputTypeStr = output.render(this.typeRenderer);
    return dedent(`
      fn ${challengeIdent}(input: ${inputTypeStr}) -> ${outputTypeStr} {
        todo!()
      }
    `);
  }

  private serialiseLinkedList(
    output: Type<Output>,
    boundary: IBoundary
  ): string {
    if (!(output.inner instanceof LinkedList)) {
      throw new Error(
        `Attempted to serialise a type that wasn't a LinkedList!`
      );
    }

    return dedent(`{
      "${boundary.marker}_linkedList": {
        "inner": result,
      }
    }`);
  }

  private serialiseDict(output: Type<Output>): string {
    if (!(output.inner instanceof Dictionary)) {
      throw new Error(
        `Attempted to serialise a type that wasn't a Dictionary!`
      );
    }

    // map the HashMap into a Vec<(key, value)> so keys aren't cast to strings when JSON encoded
    const t1 = this.typeRenderer.single(output.inner.type.key);
    const t2 = this.typeRenderer.single(output.inner.type.value);
    return `result.into_iter().collect::<Vec<(${t1}, ${t2})>>()`;
  }

  public createRunner(
    challenge: Challenge<Input, Output>,
    boundary: IBoundary,
    userCode: string
  ): string {
    const challengeIdent = this.identifier(challenge.name);
    const inputIdent = `_${this.identifier(`${boundary.marker}_input`)}`;

    // NOTE: inputs will always be the same type
    const inputTypeStr = challenge.inputs[0].render(this.typeRenderer);
    const inputValueStr = challenge.inputs
      .map((input) => input.render(this.valueRenderer))
      .join(",\n");

    const outputTypeStr = challenge.output.render(this.typeRenderer);
    return dedent(`
      ${userCode}

      fn main() {
        let ${inputIdent}: Vec<${inputTypeStr}> = vec![
          ${inputValueStr}
        ];
        for tc in ${inputIdent} {
          let json = {
            let result: ${outputTypeStr} = ${challengeIdent}(tc);
            serde_json::json!(${
              challenge.output.isDictionary()
                ? this.serialiseDict(challenge.output)
                : challenge.output.isLinkedList()
                ? this.serialiseLinkedList(challenge.output, boundary)
                : "result"
            })
          };

          println!(
            "{boundary_start}{result}{boundary_end}",
            boundary_start = "${boundary.start}",
            boundary_end = "${boundary.end}",
            result = json
          );
        }
      }
    `);
  }
}
