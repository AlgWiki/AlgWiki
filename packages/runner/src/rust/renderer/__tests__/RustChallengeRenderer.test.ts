import { IBoundary } from "../../../Boundary";
import { Challenge } from "../../../Challenge";
import { Primitive, Type } from "../../../Type";
import { RustChallengeRenderer } from "../RustChallengeRenderer";

// TODO: tests for throwIfNeedsValue

describe("RustChallengeRenderer", () => {
  const challenge = new Challenge(
    "parseStr",
    [1, 2, 3].map((x) => Type.single(Primitive.String, x.toString())),
    Type.single(Primitive.Integer)
  );

  it("can render default input", () => {
    const defaultCode = new RustChallengeRenderer().createDefaultCode(
      challenge
    );
    expect(defaultCode).toMatchSnapshot();
  });

  it("can render challenge", () => {
    const userCode = "Hello, User!";
    const boundary: IBoundary = {
      marker: "B_",
      start: "B_START",
      end: "B_END",
      error: "B_ERROR",
    };
    const runnerCode = new RustChallengeRenderer().createRunner(
      challenge,
      boundary,
      userCode
    );
    expect(runnerCode).toMatchSnapshot();
  });
});
