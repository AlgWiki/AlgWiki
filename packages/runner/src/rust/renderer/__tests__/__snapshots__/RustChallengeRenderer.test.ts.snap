// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RustChallengeRenderer can render challenge 1`] = `
"
      Hello, User!

      fn main() {
        let _b_input: Vec<&str> = vec![
          r###\\"1\\"###,
r###\\"2\\"###,
r###\\"3\\"###
        ];
        for tc in _b_input {
          let json = {
            let result: i64 = parse_str(tc);
            serde_json::json!(result)
          };

          println!(
            \\"{boundary_start}{result:?}{boundary_end}\\",
            boundary_start = \\"B_START\\",
            boundary_end = \\"B_END\\",
            result = json.to_string()
          );
        }
      }
    "
`;

exports[`RustChallengeRenderer can render default input 1`] = `
"
      fn parse_str(input: &str) -> i64 {
        todo!()
      }
    "
`;

exports[`RustChallengeRenderer can serialise a dictionary 1`] = `
"
      USER_CODE_HERE

      fn main() {
        let _b_input: Vec<HashMap<i64, i64>> = vec![
          {
      use std::collections::HashMap;
      let mut inner: HashMap<i64, i64> = HashMap::new();
      inner.insert(1, 1);
inner.insert(2, 2);
inner.insert(3, 3);
      inner
    }
        ];
        for tc in _b_input {
          let json = {
            let result: HashMap<i64, i64> = a_dictionary(tc);
            serde_json::json!(result.into_iter().collect::<Vec<(i64, i64)>>())
          };

          println!(
            \\"{boundary_start}{result:?}{boundary_end}\\",
            boundary_start = \\"B_START\\",
            boundary_end = \\"B_END\\",
            result = json.to_string()
          );
        }
      }
    "
`;

exports[`RustChallengeRenderer can serialise a linked list 1`] = `
"
      USER_CODE_HERE

      fn main() {
        let _b_input: Vec<LinkedList<i64>> = vec![
          {
      use std::collections::LinkedList;
      let mut inner: LinkedList<i64> = LinkedList::new();
      inner.push_back(1);
inner.push_back(2);
inner.push_back(3);
      inner
    }
        ];
        for tc in _b_input {
          let json = {
            let result: LinkedList<i64> = a_linked_list(tc);
            serde_json::json!({
      \\"B__linkedList\\": {
        \\"inner\\": result,
      }
    })
          };

          println!(
            \\"{boundary_start}{result:?}{boundary_end}\\",
            boundary_start = \\"B_START\\",
            boundary_end = \\"B_END\\",
            result = json.to_string()
          );
        }
      }
    "
`;
